Support for OAUTH Token for private repos

vipulchhabra99 opened the pull request:
Fixes #728 .

vipulchhabra99 commented:
@ivotron I have added the support for it in the file specified by you in the issue but It doesn't seems feasible to add the test for it.

JayjeetAtGithub commented:
@vipulchhabra99 Can you please tell why it's not feasible to add test?

JayjeetAtGithub commented:
@vipulchhabra99 We don't want to involve `.travis.yml` in storing any API key/secret. Please look at [this](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings). This way we do not need to encrypt the keys ourselves and update the `.travis.yml` multiple times.

vipulchhabra99 commented:
Hi @JayjeetAtGithub , This link points to defining the environment variables in Travis settings of the repo and I think I don't have the rights to add it in environment variables.

ivotron commented on cli/popper/scm.py:
```diff
@@ -183,6 +183,12 @@ def get_default_branch(r):
 
+        auth_token = os.getenv('GITHUB_API_TOKEN')
+
+        if(auth_token is not None):
```
can we add logic for checking that the URL is pointing to github? If the URL is pointing somewhere else it'll likely fail.

vipulchhabra99 replied:
Yeah, sure I can add it.

ivotron commented on cli/popper/scm.py:
```diff
@@ -183,6 +183,12 @@ def get_default_branch(r):
+        auth_token = os.getenv('GITHUB_API_TOKEN')
+
+        if(auth_token is not None):
+            if('@' not in url):
```
this conditional will result in not being able to work with URLs of the form `git@github.com`, which is OK but we need to add this to the documentation.

vipulchhabra99 replied:
Yes, I can add this to documentation to make it clear to the user.

ivotron commented on cli/popper/scm.py:
```diff
@@ -183,6 +183,12 @@ def get_default_branch(r):
+
+        if(auth_token is not None):
+            if('@' not in url):
+                url = url[:8]+auth_token+'@'+url[8:]
```
can you add a comment explaining why 8 characters before and after`url[:8]`?

vipulchhabra99 replied:
Yeah sure.

vipulchhabra99 replied:
Actually I just realized that there can be other protocols defined http:// and git@ although git@ would automatically wouldn't be considered for adding token, I can add support for it too.

vipulchhabra99 replied:
@ivotron Should I also add a check that protocol should be HTTPS only because Token works only with HTTPS protocol.

ivotron commented on .travis.yml:
```diff
@@ -1,11 +1,13 @@
+  - RUNTIME=singularity POPPER_TEST_MODE=with-git
+  - RUNTIME=docker POPPER_TEST_MODE=without-git
+  global:
+  - secure: DXDmJmyxb89CZOI4DEoPQE4iI+tna3TDwf0UekgIt36GqJfqDpYjtaar6tZVWhDZVdswOxxe+ulnRnSLLb0l4EDuPnOPu6oiBo9i8q1t5qy1bsaiyyM9pkOnZbKcNEvF9cSv15a/A/m1y6/aOSWFJDaSnKWUu1u0sQrUGuY7XlZCy9lns1p8diAeatQMx9tBy5bcJi4L6xcrHwNMs6IJnBcYEAxKAKHkNMGUOPObVnOyGGLfo4T4hTzVouw7i/wZHfS1KIM/KPYOZuZfNVK/NgV3m4YKlV2Y+zj8T1fsQH/aEIkEoQwUGbZhPofvZIncJ7v2z7TC2qW76DfuwOslDPkxqZ7Lw2/Wx22T3NgKsFKKwFaseH1p+M/PhgS5APR6qU/ ...
```
this can be removed

vipulchhabra99 replied:
OK I can remove this

ivotron commented:
@vipulchhabra99 I've added a `GITHUB_API_TOKEN` variable to travis associated to my account. If you create a private action/repo and give my github user access, our travis will be able to test these feature. Note that the test of the PR is not going to pick this env variable up, as workflows triggered by PRs cannot read environment variables. What we can do is that you can test locally on your machine, and once everything works on your end you can post here and I'll manually trigger on this PR. Would that work?

pep8speaks commented:
Hello @vipulchhabra99! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cli/popper/scm.py`](https://github.com/systemslab/popper/blob/30fd8f9df72fafb648367d8b2bc81b156b656ba4/cli/popper/scm.py):

> [Line 186:9](https://github.com/systemslab/popper/blob/30fd8f9df72fafb648367d8b2bc81b156b656ba4/cli/popper/scm.py#L186): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 192:21](https://github.com/systemslab/popper/blob/30fd8f9df72fafb648367d8b2bc81b156b656ba4/cli/popper/scm.py#L192): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 193:21](https://github.com/systemslab/popper/blob/30fd8f9df72fafb648367d8b2bc81b156b656ba4/cli/popper/scm.py#L193): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 193:80](https://github.com/systemslab/popper/blob/30fd8f9df72fafb648367d8b2bc81b156b656ba4/cli/popper/scm.py#L193): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)

vipulchhabra99 commented:
Hi, @ivotron I have updated the pr as per your review please check them once. Also, I have added you as a collaborator in that dummy repo from my end please accept invitation so that you can access that private repository and also I have one question that If we have somehow passed the integration tests for it but later on if someone else sends new pull request to the repo will this test would be triggered again? If triggered then I think then Travis would result for build failed for every pull request.

ivotron commented:
@vipulchhabra99 you're right. An alternative is to check for whether `GITHUB_TOKEN_API` variable is defined and skip the test if is not.

ivotron commented on .travis.yml:
```diff
@@ -1,11 +1,10 @@
 env:
-- RUNTIME=singularity POPPER_TEST_MODE=with-git
-- RUNTIME=docker POPPER_TEST_MODE=without-git
+  - RUNTIME=singularity POPPER_TEST_MODE=with-git
```
since these are cosmetic changes, we can leave them out of the PR

ivotron commented on cli/popper/scm.py:
```diff
@@ -232,6 +232,18 @@ def get_default_branch(r):
+
+        if(auth_token is not None):
+            # To verify the link of github for private repo support.
+            if('github' in url):
```
this can be a single `if` statement (conjunction of the three `if` conditions)

ivotron commented on cli/popper/scm.py:
```diff
@@ -232,6 +232,18 @@ def get_default_branch(r):
+                if('@' not in url):
+                    # The authentication token has to be added after protocol
+                    # The length of protocol is 8 in case of https://
+                    if('https' in url):
```
HTTP is the only protocol used by Popper, so `https` is always in the URL

ivotron commented on docs/sections/gha_workflows.md:
```diff
@@ -187,6 +187,14 @@ refer to an action on a public Git repository or Docker container registry:
 
 -->
 
+### Using Private repositories of `Github` in an action
```
This can be `### Referencing private Github repositories in an action`

ivotron commented on docs/sections/gha_workflows.md:
```diff
@@ -187,6 +187,14 @@ refer to an action on a public Git repository or Docker container registry:
 
+### Using Private repositories of `Github` in an action
+
+When creating an action block we can also use an action defined in Github private repository. To use private repository you need to create Personal access token from <a href="https://github.com/settings/tokens"> here </a> with permissions to access private repositories and set it as an environment variable using
```
couple of changes (note that readthedocs accepts markdown):

We can make use of actions located in private Github repositories by defining a `GITHUB_API_TOKEN` environment variable that the `popper run` command can inspect. To accomplish this, the repository referenced in the `uses` attribute is assumed to be private and, to access it, an API token from Github is needed (see instructions [here](https://github.com/settings/tokens)). The token needs to have permission to read the private repository in question. To define the `GITHUB_API_TOKEN` variable:

```
export GITHUB_API_TOKEN=access_token_here
popper run
```

If the access token doesn't have permissions to access private repositories, the `popper run` command will fail.

vipulchhabra99 commented:
Hi @ivotron, Thanks a lot for reviewing the pr I have updated it as per your review. Please have a look once

ivotron commented:
thanks a lot @vipulchhabra99 !

vipulchhabra99 commented:
Thanks, a lot @ivotron and @JayjeetAtGithub, Your guidance helped me a lot to solve the issue.
